dynamic_instrumentation/src/main.rs

```diff
@@ -0,0 +1,383 @@
+                    "/dev/stdin".into()
+                } else {
+                    fs_err::canonicalize(src_path)?
+                }
+            }
+            free_matches => {
+                return Err(eyre!(
+                    "`rustc` args `matches.free` is not a single source path: {free_matches:?}"
+                ))
+            }
+        };
+        let Options {
+            crate_name,
+            crate_types,
+            edition,
+            ..
+        } = session_options;
+        let crate_name = crate_name.ok_or_else(|| eyre!("no crate_name specified by `cargo`"))?;
+        Ok(Self {
+            crate_name: crate_name.as_str().into(),
+            src_path,
+            crate_types: crate_types.into_iter().map(CrateType::from).collect(),
+            edition: edition.into(),
+        })
+    }
+
+    pub fn from_metadata(metadata: &Metadata) -> eyre::Result<Vec<Self>> {
+        Self::from_metadata_package(
+            metadata
+                .root_package()
+                .ok_or_else(|| eyre!("no root package found by `cargo`"))?,
+        )
+    }
+
+    pub fn from_rustc_args(at_args: &[String]) -> eyre::Result<Self> {
+        let args = rustc_driver::args::arg_expand_all(at_args);
+        let matches = rustc_driver::handle_options(&args)
+            .ok_or_else(|| eyre!("failed to parse `rustc` args"))?;
+        let session_options = rustc_session::config::build_session_options(&matches);
+        let crate_target = CrateTarget::from_session_options(session_options, &matches.free)?;
+        Ok(crate_target)
+    }
+}
+
+#[derive(Debug, Serialize, Deserialize)]
+struct InstrumentInfo {
+    crate_targets: Vec<CrateTarget>,
+    metadata: PathBuf,
+}
+
+fn get_sysroot_fast() -> Option<PathBuf> {
+    let sysroot = [
+        env::var_os("RUSTUP_HOME")?,
+        "toolchains".into(),
+        env::var_os("RUSTUP_TOOLCHAIN")?,
+    ]
+    .into_iter()
+    .collect();
+    Some(sysroot)
+}
+
+fn get_sysroot_slow() -> eyre::Result<PathBuf> {
+    todo!("use `rustc --print sysroot` to support non-`rustup` cases, which @fw-immunant uses")
+}
+
+fn main() -> eyre::Result<()> {
+    color_eyre::install()?;
+    env_logger::init();
+
+    let rustc_wrapper_var = "RUSTC_WRAPPER";
+    let instrument_info_var = "C2RUST_INSTRUMENT_INFO";
+
+    let own_exe = env::current_exe()?;
+    let own_exe = own_exe.as_path();
+
+    let wrapping_rustc = env::var_os(rustc_wrapper_var).as_deref() == Some(own_exe.as_os_str());
+    if wrapping_rustc {
+        let sysroot = get_sysroot_fast()
+            .ok_or(())
+            .or_else(|()| get_sysroot_slow())?;
+        let mut at_args = env::args().skip(1).collect::<Vec<_>>();
+        let crate_target = CrateTarget::from_rustc_args(&at_args)?;
+        let info = env::var(instrument_info_var)?;
+        let info = serde_json::from_str::<InstrumentInfo>(&info)?;
+        let should_instrument = info.crate_targets.contains(&crate_target);
+        if !should_instrument {
```

Comment:
nit: put the `should_instrument` path first for style

Reply:
I did it this way since I was going to skip the `else`, as this first branch is shorter and ends with an `exit_with_status`, which returns `!`, but that's unstable, so I can't mark it as such.  I'm probably gonna end up refactoring things more so this might change anyways.

Reply:
Fixed in https://github.com/immunant/c2rust/pull/554/commits/b6c98aca1e19cb5e03f9b071dd0ce13256a36c82.